src/uct/base/uct_iface.h

```diff
@@ -396,6 +396,110 @@ typedef struct uct_iface_local_addr_ns {
     }
 
 
+/**
+ * Call TL registration routine.
+ *
+ * @param [in] _name  TL name
+ */
+#define uct_tl_register(_name) \
+    uct_tl_##_name##_ctor()
+
+
+/**
+ * Call TL unregistration routine.
+ *
+ * @param [in] _name  TL name
+ */
+#define uct_tl_unregister(_name) \
+    uct_tl_##_name##_dtor()
+
+
+/**
+ * Register component and TL
+ *
+ * @param [in] _name  Component and TL name
+ * @param [in] ...    Transport and component registration routines
+ *                    attributes (static)
+ */
+#define UCT_COMP_TL_REGISTER(_name, ...) \
+    UCT_COMPONENT_REGISTER_DECL(_name, __VA_ARGS__); \
+    UCT_TL_REGISTER_DECL(_name, __VA_ARGS__); \
+    static void uct_component_tl_##_name##_ctor(void); \
+    void UCS_F_CTOR uct_##_name##_init(void) { \
+        uct_component_tl_##_name##_ctor(); \
+        uct_component_register(_name); \
+        uct_tl_register(_name); \
+    } \
+    void uct_component_tl_##_name##_ctor(void) \
+
+
+/**
+ * Unregister component and TL
+ *
+ * @param [in] _name  Component and TL name
+ * @param [in] ...    Transport and component registration routines
+ *                    attributes (static)
+ */
+#define UCT_COMP_TL_UNREGISTER(_name, ...) \
+    UCT_COMPONENT_REGISTER_DECL(_name, __VA_ARGS__); \
+    UCT_TL_REGISTER_DECL(_name, __VA_ARGS__); \
```

Comment:
1. maybe remove UCT_TL_REGISTER_DECL, and here we just
`__VA_ARGS__ void uct_tl_##_name##_dtor(void);`
similarly for the ctor

2. why the function attributes cannot be one macro argument, instead of __VA_ARGS__?

Reply:
> * maybe remove UCT_TL_REGISTER_DECL, and here we just
>   `__VA_ARGS__ void uct_tl_##_name##_dtor(void);`
>   similarly for the ctor

ok, done

> * why the function attributes cannot be one macro argument, instead of **VA_ARGS**?

one argument will always require additional argument else we get compilation macro, VA_ARGS may just skip argument. use mandatory argument instead?